Feature/planning comparison

mugwump opened the pull request:
Restoring historical values for a given set of filtered work_packages.

mugwump committed:
added a planning-comparison-service

The PlanningComparisonService restores tthe work_packages at a given time
from the latest journal-entries for the work_package. This commit
contains the sql to restore the work_packages and includes a simple
test to make sure, that the returned work_packages actually contain the
old attribute-values.

mugwump committed:
adding an additional journal-entry

this makes sure, that the journalization actually picks up the right
journal-entry.

mugwump committed:
Added the filters to planning-comparison

mugwump committed:
added the compare_date to the planning_elements_controller

The planning-elements-controller now accepts a historical date to
return work_packages at a given time.

There is now a cuke, that checks, that the api is actually returning
the historical values; there are also some general steps to use timecop
for producing historical data.

mugwump committed:
added the filtering to the cuke

mugwump committed:
replaced compare_date with at_time

mugwump committed:
fixing the current-workpackages: Historical workpackages can stil only be obtained for only one project

mugwump committed:
Now searching in multiple projects for historical data is possible

mugwump committed:
Added a cuke for the due_date-comparison; made the api more robust against missing filters.

mugwump committed:
work_packages are now resolved correctly when no filter is given

mugwump committed:
using DateTime instead of date to parse the give at_time

mugwump committed:
Moved the timecop-steps aout of the api into the timecop-steps

apfelfabrik committed:
Enables comparisons for timelines.

mugwump committed:
added a missing reset after timetravel

mugwump committed:
removed the Proc.new around the timecop-cleanup: now the time is actually reset after each run.

apfelfabrik committed:
Enables client-side fetching of historical data.

mugwump committed:
making the planning-comparison-sql run on mysql.

mugwump committed:
now accepting only timestamps/seconds from epoch

mugwump committed:
added some puts to find out wherre exactly the comparison fails on ci

mugwump committed:
changed the work_package-select: We're looking for the journal_id, not for the id of the work_package_journal.

apfelfabrik committed:
Refactors query string builder.

apfelfabrik committed:
Timelines now use the new query string builder API.

mugwump committed:
removed some debugging-statements

mugwump committed:
Merge branch 'feature/planning_comparison' of github.com:opf/openproject into feature/planning_comparison

mugwump committed:
fixed the wrong select of the latest journal-entries

mugwump committed:
the api accepts the at_time only as unix-timestamp

apfelfabrik commented:
Types, Statuses and Projects should be prefetched, as those trigger subsequent queries to the database, per work package that is included in the comparison.

apfelfabrik commented:
Also, something triggers queries of the form

```
SELECT COUNT(*) FROM "work_packages" WHERE "work_packages"."parent_id" = 99
```

to the database, apparently also per work package that is included in the comparison.

mugwump committed:
references are now filled without new queries
not sure, if this is really effective, but it replaces the n+1 calls
with significantly less calls. Still need to step through all
workpackages several times to gather&fill all relations.

There is still a parent.children.size-query left, that needs to be
killed.

Either that, or get rid of these resolved&redundant references
alltogether and replace them with simple id's (e.g use project_id =1
instead of the redunedant project: {id: 1, name: "seeded project" ...}

mugwump committed:
some optimisations to get around unnecessary db-calls.

apfelfabrik committed:
Adds cukes for timeline comparisons.

apfelfabrik committed:
Fixes cuke tables and indentation.

apfelfabrik committed:
Moves planning element steps into timelines steps.

apfelfabrik committed:
Removes planning element steps.

apfelfabrik committed:
Implements time-travelling background for comparison cukes.

apfelfabrik committed:
Merge branch 'feature/planning_comparison' of github.com:opf/openproject into feature/planning_comparison

apfelfabrik committed:
Supports parts of the scenarios.

mugwump committed:
fixes an issue where broken journals with exactly the same
dates/timestamps are returned twice: the filter then returned two
identical work_packages. This adds the max(id) to the journal-select,
which will then only return one of the identical work-packages.

apfelfabrik committed:
Merge branch 'feature/rails3' into feature/planning_comparison

apfelfabrik committed:
Adds more steps and rights to the comparison cukes.

apfelfabrik committed:
Disables cukes that have timecop problems.

apfelfabrik committed:
Changelog.

apfelfabrik commented:
```
Task #1880: Re-introduce at-time scope
Task #1881: Re-introduce project planning comparison in controller
Task #1994: Integrational tests for work packages at_time (API)
Task #2173: Adapt client-side to new server behavior
Task #1883: Extend at-time scope for status comparison
Task #1884: Make status values available over API
```

apfelfabrik committed:
Merge remote-tracking branch 'opf/feature/rails3' into feature/planning_comparison

Conflicts:
	doc/CHANGELOG.md

hschink commented on app/services/planning_comparison_service.rb:
```diff
@@ -0,0 +1,101 @@
+
+    def self.resolve_reference_ids(work_packages)
+      # TODO faster ways to do this without stepping numerous times through the workpackages?!
+      # Or simply wait unti we finally throw out the redundant references out of the json/xml-rendering??!
```
Just a typo: `unti` should be `until` I guess...

hschink commented on features/timelines/timeline_comparison_view.feature.disabled:
```diff
@@ -0,0 +1,81 @@
+#-- copyright
```
Why is this file deactivated?

apfelfabrik replied:
It is still in progress in `#1993`.

apfelfabrik committed:
Fixes typo.

apfelfabrik committed:
Merge remote-tracking branch 'opf/feature/rails3' into feature/planning_comparison

Conflicts:
	doc/CHANGELOG.md
